### PR TITLE
Enable automated Prune via CRD NodeFeatureDiscoveries

### PR DESCRIPTION
**Pull request by ArangoGutierrez**
Fixes: #187

### CODE REVIEW

File: api/v1/nodefeaturediscovery_types.go [hunk 1]
```diff
@@ -63,6 +63,12 @@ type NodeFeatureDiscoverySpec struct {
 	// worker.
 	// +optional
 	WorkerConfig ConfigMap `json:"workerConfig"`
+
+	// PruneOnDelete defines whether the NFD-master prune should be
+	// enabled or not. If enabled, the Operator will deploy an NFD-Master prune
+	// job that will remove all NFD labels from the cluster nodes.
```
Comment (marquiz) [inline api/v1/nodefeaturediscovery_types.go:69]:
```suggestion
	// job that will remove all NFD labels (and other NFD-managed assets such as annotations, extended resources and taints) from the cluster nodes.
```

**Review by marquiz — COMMENTED** [review]
Thanks @ArangoGutierrez. I'll take a closer look in the coming days

File: controllers/nodefeaturediscovery_controls.go [hunk 2]
```diff
@@ -528,6 +529,67 @@ func Deployment(n NFD) (ResourceStatus, error) {
...
+	// SetControllerReference sets the owner as a Controller OwnerReference
+	// and is used for garbage collection of the controlled object. It is
+	// also used to reconcile the owner object on changes to the controlled
+	// object. If we cannot set the owner, then return NotReady
+	if err := controllerutil.SetControllerReference(n.ins, &obj, n.rec.Scheme); err != nil {
+		return NotReady, err
+	}
+
+	// Look for the Job to see if it exists, and if so, check if it's
+	// Ready/NotReady. If the DaemonSet does not exist, then attempt to
```
Comment (marquiz) [inline controllers/nodefeaturediscovery_controls.go:568]: DaemonSet??

File: controllers/nodefeaturediscovery_controls.go [hunk 3]
```diff
@@ -528,6 +529,67 @@ func Deployment(n NFD) (ResourceStatus, error) {
...
+		if err != nil {
+			klog.ErrorS(err, "Couldn't create Job", "name", obj.Name, "namespace", obj.Namespace)
+			return NotReady, err
+		}
+		return Ready, nil
+	} else if err != nil {
+		return NotReady, err
+	}
+
+	// If we found the Job, let's attempt to update it
```
Comment (marquiz) [inline controllers/nodefeaturediscovery_controls.go:583]: Hmm, is it generally useful to update the job if it happens to exist? It doesn't retrigger it, if I'm not mistaken. Should we just delete or re-create or what would be the right thing to do?

File: controllers/nodefeaturediscovery_finalizers.go [hunk 4]
```diff
@@ -363,6 +373,101 @@ func (r *NodeFeatureDiscoveryReconciler) doComponentsExist(ctx context.Context,
...
+			return false, interpretError(err, "Prune ClusterRoleBinding")
+		}
+		klog.Info("nfd-prune ClusterRoleBinding resource has been deleted.")
+		return true, nil
+	})
+	if err != nil {
+		return err
+	}
+
+	// Attempt to delete the Job
```
Comment (marquiz) [inline controllers/nodefeaturediscovery_finalizers.go:455]: hit: shouldn't the job basically be delete before the RBAC rules as it kinda depends on them

**Review by marquiz — COMMENTED** [review]
Thanks @ArangoGutierrez for this enhancement. From what I can tell, looks pretty legit to me (if it works, haven't tested 😅). Just a few comments but nothing big

File: controllers/nodefeaturediscovery_finalizers.go (hunk 4 above)
Comment (ArangoGutierrez) [reply controllers/nodefeaturediscovery_finalizers.go:455]: jeje yeah makes sense

File: build/assets/prune/0200_clusterrole.yaml [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+- apiGroups:
+  - ""
+  resources:
+  - nodes
+  verbs:
+  - get
+  - patch
+  - update
+  - list
+  - list
```
[the remark this answers is not in the document]
Comment (ArangoGutierrez) [reply build/assets/prune/0200_clusterrole.yaml:15]: https://github.com/kubernetes-sigs/node-feature-discovery/blob/master/deployment/overlays/prune/master-clusterrole.yaml

File: build/assets/prune/0200_clusterrole.yaml [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+- apiGroups:
+  - nfd.k8s-sigs.io
+  resources:
+  - nodefeatures
+  - nodefeaturerules
+  verbs:
+  - get
+  - list
+  - watch
```
[the remark this answers is not in the document]
Comment (ArangoGutierrez) [reply build/assets/prune/0200_clusterrole.yaml:24]: https://github.com/kubernetes-sigs/node-feature-discovery/blob/master/deployment/overlays/prune/kustomization.yaml#L14

[the remark this answers is not in the document]
Comment (marquiz) [reply build/assets/prune/0200_clusterrole.yaml:24]: Oh, feel free to fix that (it's been rendered unnecessary) 👍😅

File: controllers/nodefeaturediscovery_controls.go (hunk 3 above)
Comment (ArangoGutierrez) [reply controllers/nodefeaturediscovery_controls.go:583]: good catch.
first check if successful and if not do something else

**Review by marquiz — APPROVED** [review]
Thanks @ArangoGutierrez. Looks like we're done here
/lgtm

**Pull request merged**